src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.cs

```diff
@@ -180,34 +181,41 @@ public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? optio
 
         private void CreateNodes()
         {
-            if (_list == null)
+            // Even though _list initialization can be subject to races,
+            // ensure that contending threads use a coherent view of jsonElement.
+
+            JsonElement? jsonElement = _jsonElement;
```

Comment:
`JsonElement` is a struct and cannot be read atomically.

Reply:
I don't see how the use of Interlocked.MemoryBarrier addresses that.  One thread can be doing `_jsonElement = null` while another is doing `JsonElement? jsonElement = _jsonElement;` and the read/write can be torn.

Reply:
Is the idea that a thread will only null out _jsonElement after setting _list to be not null, and a thread will only care about the read value of _jsonElement if it subsequently observed a null list?

Reply:
Correct, as long as _list is null torn reads would be avoided.

Reply:
I think it's more that there may be torn reads, but such tearing will be benign if the _list is non-null.

A more explicit comment would be helpful.

Reply:
I used MemoryBarrier because I'm not aware of a different way to arrange non-atomic read/writes.

Reply:
Added comments.